### PR TITLE
Ensure quantity inputs update if receiving new props from server

### PR DESCRIPTION
**Pull request by opr**
### Changes proposed in this Pull Request:

When updating a product quantity on the client, the quantity input component tracks the expected next value when increasing/decreasing or changing the value in the input field (e.g. if you are on 9, and press + the expected next value is 10). If the server updates the quantity prop in an unexpected way, then it is ignored, because this likely means the shopper is in the middle of changing the value. The method by which the shopper is updating (increase, decrease, or input) is also tracked, specifically so we can avoid overwriting shoppers' inputs while in the middle of typing a quantity into the box.

If the _updated_ quantity (which comes from the server) matches the expected quantity, then the value is applied and the expected "method" is reset.

There was a bug where the "method" of this update was not cleared, so valid server updates were not applied to the input field.

This PR ensures that expected method is reset after successful quantity application to allow updates from the prop to apply again.

Closes [WOOPLUG-4949](https://linear.app/a8c/issue/WOOPLUG-4949/quantity-inputs-do-not-update-with-new-props-from-server)

(For Bug Fixes) Bug introduced in PR https://github.com/woocommerce/woocommerce/pull/58693

### Screenshots or screen recordings:

| Before | After |
| ------ | ----- |
| https://www.loom.com/share/42e1e0dfeec844aea00ddc76c20d2948 | https://www.loom.com/share/997183ab5e6741189e4bece0f42600fb |

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure your have product bundled installed along with the changes of latest trunk.
2. Create a bundled product with a child product, and set the individual price of the child product. Do not set the price for the parent product for simplicity.
3. Add that bundled product to the cart. Goto cart.
4. Increase the quantity of the parent item, you'll see the child item quantity updates as expected.
5. Now reduce the quantity of the child item.
6. Try again to increase the quantity of the parent item, ensure the child item quantity updates as expected.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Review by PanosSynetos — APPROVED** [review]
Manually tested with Product Bundles and it works as expected, the reported issue is resolved. - I'll let the code review to Rubik, that has more experience on the block.

# Testing

Observe that in `trunk` when I reduce the quantity of `Rush Downhill Speed Bombers Hybrid` and then increase the bundle parent quantity, the quantity of `Rush Downhill` doesn't get updated.

With this branch, if I do the same as above the quantity of `Rush Downhill` is updated correctly

## Trunk

https://github.com/user-attachments/assets/8bc5ffb7-4e3e-4aa7-9deb-fec6c0f7dc90

## This Branch
https://github.com/user-attachments/assets/bbac7f71-2e97-4ca6-a28b-92f9dc727778

**Review by naman03malhotra — APPROVED** [review]
Thanks Robert for the quick fix.

p1751871843800789-slack-CNA89KY6M e2e failure was very notorious and difficult to find as it was only happening for an edge case, when clicked on the -ve button to decrease the quantity.

[Detailed context](p1751871843800789-slack-CNA89KY6M)

I followed these steps to test the issue:

1. **Created** a bundled product with a single child product and set an individual price for the child. I left the parent product price unset for simplicity.  
2. **Added** the bundled product to the cart.  
3. **Increased** the quantity of the parent item on the cart page — the child item quantity updated as expected.  
4. **Reduced** the quantity of the child item manually.  
5. **Increased** the parent item quantity again, the child items updates.

E2E tests are passing again.

image

File: plugins/woocommerce/client/blocks/assets/js/base/components/quantity-selector/index.tsx [hunk 1]
```diff
@@ -115,6 +115,8 @@ const QuantitySelector = ( {
 	 */
 	useLayoutEffect( () => {
```
Comment (ralucaStan) [inline plugins/woocommerce/client/blocks/assets/js/base/components/quantity-selector/index.tsx:116]: not related to this PR but I wonder why we are doing useLayoutEffect here, we're not doing layout calculations here

**Review by ralucaStan — APPROVED** [review]
I tested this following the testing steps in:
- the Cart block and it works as expected; I also tested it with low stock products in the bundle; that blocks the + sign, which I expect it's an expected outcome
- the Mini Cart block - the functionality works, but the layout isn't good, see screenshot below. Not related to this PR though
Screenshot 2025-07-11 at 15 05 39
- in the Checkout block with a Product collection on the same page. @PanosSynetos shoppers can use a product collection block to add to cart; if I add a product that is in a bundle, it is added inside the bundle. see the video below, it might be something you want to also test separately as it's not related to this PR

https://github.com/user-attachments/assets/956484e6-9008-44e1-88ad-f39c123340da

**Comment by jimjasson** [comment]
Hey @ralucaStan, 

Can you please provide some more details about the issue with Bundles that you discovered? The expected behavior here would be for the product to be added as a separate line item in the cart, and not as part of the Bundle as it wont have the same cart item key. 

What is the block under the Cart block? Is it a Product Collection block? Do you see the same issue if you add the product to the cart from the single product page?

I tried replicating this issue using this branch and a Product Collection block with handpicked items, but the product from the collection was added as a separate cart item as expected:

[Screen Capture on 2025-07-14 at 13-41-25.mp4](https://uploads.linear.app/f629fcde-6f9b-48e8-8af3-60eed8823bff/0aff7222-87fb-42af-b88a-d13c981eb1c5/b9c5156b-fc17-4030-a555-1bd1c4822670?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiL2Y2MjlmY2RlLTZmOWItNDhlOC04YWYzLTYwZWVkODgyM2JmZi8wYWZmNzIyMi04N2ZiLTQyYWYtYjg4YS1kMTNjOTgxZWIxYzUvYjljNTE1NmItZmMxNy00MDMwLWE1NTUtMWJkMWM0ODIyNjcwIiwiaWF0IjoxNzUyNDg5NzkyLCJleHAiOjMzMzIzMDQ5NzkyfQ.Na-CDVTSCkGY75vhj9lZj7PmbGCgstEM6TiKcWc-nLY)

File: plugins/woocommerce/client/blocks/assets/js/base/components/quantity-selector/index.tsx (hunk 1 above)
Comment (opr) [reply plugins/woocommerce/client/blocks/assets/js/base/components/quantity-selector/index.tsx:116]: `useLayoutEffect` runs synchronously before the browser repaints, ensuring any quantity updates from the server happen before rendering to prevent flickering if the quantity changes rapidly. It doesn't have to relate to layout calculations, it's just to do something before the repaint step.
